File: 1D Array/C/program_12.c
#include <stdio.h>
#include <stdlib.h>


typedef struct {
    int key;
    int value;
} HashNode;

#define TABLE_SIZE 1000

HashNode hashTable[TABLE_SIZE];


int hash(int key) {
    return abs(key) % TABLE_SIZE;
}


void insert(int key, int value) {
    int index = hash(key);
    while (hashTable[index].key != 0) {  
        index = (index + 1) % TABLE_SIZE;
    }
    hashTable[index].key = key;
    hashTable[index].value = value;
}


int search(int key) {
    int index = hash(key);
    while (hashTable[index].key != 0) {
        if (hashTable[index].key == key) {
            return hashTable[index].value;
        }
        index = (index + 1) % TABLE_SIZE;
    }
    return -1; 
}


void twoSum(int nums[], int n, int target) {
    
    for (int i = 0; i < TABLE_SIZE; i++) {
        hashTable[i].key = 0;
    }

    for (int i = 0; i < n; i++) {
        int complement = target - nums[i];

        int index = search(complement);
        if (index != -1) {
            printf("[%d, %d]\n", index, i);
            return;
        }

        insert(nums[i], i);
    }

    printf("Not found\n");
}

int main() {
    int nums[] = {2, 6, 18, 19};
    int target = 8;
    int n = sizeof(nums) / sizeof(nums[0]);

    printf("elements that sum  %d:\n", target);
    twoSum(nums, n, target);

    return 0;
}

File: 1D Array/java/program12.java
import java.util.HashMap;
import java.util.Map;

public class program12 {
    public static void twoSum(int[] nums, int target) {
        Map<Integer, Integer> hashTable = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            if (hashTable.containsKey(complement)) {
                System.out.println("[" + hashTable.get(complement) + ", " + i + "]");
                return;
            }

            hashTable.put(nums[i], i);
        }

        System.out.println("Not found");
    }

    public static void main(String[] args) {
        int[] nums = {2, 6, 18, 19};
        int target = 8;

        System.out.println("Elements that sum to " + target + ":");
        twoSum(nums, target);
    }
}